crates/rpc/src/methods/token_add.rs

```diff
@@ -0,0 +1,317 @@
+use crate::{Error, Result};
+use ethui_dialogs::{Dialog, DialogMsg};
+use ethui_networks::Networks;
+use ethui_types::{Address, GlobalState, TokenMetadata, U256};
+use ethui_wallets::{WalletControl, Wallets};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub struct TokenAdd {
+    erc20_token: Option<TokenMetadata>,
+    erc721_token: Option<ERC721Data>,
+    erc1155_token: Option<ERC1155Data>,
+    chain_id: Option<u32>,
+    _type: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct ERC721Data {
+    address: Address,
+    token_id: U256,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct ERC1155Data {
+    address: Address,
+    token_id: U256,
+}
+
+impl TokenAdd {
+    pub fn build() -> TokenAddBuilder {
+        TokenAddBuilder::default()
+    }
+
+    pub async fn get_current_chain_id(&self) -> Result<u32> {
+        let networks = Networks::read().await;
+        let current_network = networks.get_current();
+        Ok(current_network.chain_id)
+    }
+
+    pub async fn get_current_wallet_address(&self) -> Result<Address> {
+        let wallets = Wallets::read().await;
+        let wallet_address = wallets.get_current_wallet().get_current_address().await;
+        Ok(wallet_address)
+    }
+
+    pub fn check_type(&self) -> Result<()> {
+        if self._type != "ERC20" && self._type != "ERC721" && self._type != "ERC1155" {
+            return Err(Error::TypeInvalid(self._type.clone()));
+        }
+
+        Ok(())
+    }
+
+    pub fn check_symbol(&self) -> Result<()> {
+        if let Some(erc20_token) = &self.erc20_token {
+            let symbol = erc20_token.symbol.clone().unwrap_or_default();
+            if symbol.len() > 11 {
+                return Err(Error::SymbolInvalid(symbol));
+            }
+        }
+        Ok(())
+    }
+
+    pub fn check_decimals(&self) -> Result<()> {
+        if let Some(erc20_token) = &self.erc20_token {
+            let decimals = erc20_token.decimals.unwrap_or(0);
+            if decimals > 36 {
+                return Err(Error::DecimalsInvalid(decimals));
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn check_network(&self) -> Result<()> {
+        let current_chain_id = self.get_current_chain_id().await?;
+        let chain_id = self.chain_id.unwrap_or(current_chain_id);
+        if current_chain_id != chain_id {
+            return Err(Error::NetworkInvalid);
+        }
+        Ok(())
+    }
+
+    pub async fn run(self) -> Result<()> {
+        self.check_type()?;
+        self.check_network().await?;
+
+        match self._type.as_str() {
+            "ERC20" => {
+                self.check_symbol()?;
+                self.check_decimals()?;
```

Comment:
these functions already do nothing if you pass in another token type. so it seems you're double filtering the token type.

and the main problem: if `self._type` is `ERC20`, but `self.erc20_token` is `None`, these checks will still go through, but they should return an error due to the value being missing.